src/toil/lib/retry.py

```diff
@@ -13,31 +13,145 @@
-standard_library.install_aliases()
-from builtins import next
-from contextlib import contextmanager
 import time
-import urllib.request, urllib.error, urllib.parse
-try:
-    from httplib import BadStatusLine
-except ImportError:
-    # This has moved in Python 3
-    from http.client import BadStatusLine
+import copy
+import functools
 import logging
 
-log = logging.getLogger( __name__ )
+from contextlib import contextmanager
+from typing import List, Set, Optional, Tuple, Callable, Any
+
+log = logging.getLogger(__name__)
+
+
+class ErrorCondition:
+    """
+    A wrapper describing an error condition.
+
+    ErrorCondition events may be used to define errors in more detail to determine whether to retry.
+
+    :param error: An Exception (required)
+    :param error_codes: Error codes that must match to be retried (optional; defaults to not checking)
+    :param error_message_must_include: A string that must be in the error message to be retried
+        (optional; defaults to not checking)
+    :param retry_on_this_condition: This can be set to False to always error on this condition.
+    """
+    def __init__(self,
+                 error: Any,
+                 error_codes: List[int] = None,
+                 error_message_must_include: str = None,
+                 retry_on_this_condition: bool = True):
+        self.error = error
+        self.error_codes = error_codes
+        self.error_message_must_include = error_message_must_include
+        self.retry_on_this_condition = retry_on_this_condition
+
+
+def retry_decorator(intervals: Optional[List] = None,
+                    infinite_retries: bool = False,
+                    errors: Optional[Set] = None,
+                    error_conditions: Optional[List[ErrorCondition]] = None,
+                    log_message: Optional[Tuple[Callable, str]] = None):
+    """
+    Retry a function if it fails with any Exception defined in the "errors" set, every x seconds,
+    where x is defined by a list of numbers (ints or floats) in "intervals".  Also accepts ErrorCondition events
+    for more detailed retry attempts.
+
+    :param Optional[List] intervals: A list of times in seconds we keep retrying until returning failure.
+        Defaults to retrying with the following exponential back-off before failing:
+            1s, 1s, 2s, 4s, 8s
+    :param infinite_retries: If this is True, reset the intervals when they run out.  Defaults to: False.
+    :param errors: A set of exceptions to catch and retry on.
+    :param error_conditions: A list of ErrorCondition objects describing more detailed error event conditions.
+        An ErrorCondition specifies:
+            - Exception (required)
+            - Error codes that must match to be retried (optional; defaults to not checking)
+            - A string that must be in the error message to be retried (optional; defaults to not checking)
+            - A bool that can be set to False to always error on this condition.
```

Comment:
They are OR-ed.  Either an error in `errors` or `error_conditions` is sufficient to retry on.

If an error is in `errors` and a more specific version of it is in `error_conditions` it will only retry on the more specific version currently.  I think that this should retry on the more general one though, and will change this.

Reply:
I've refactored it to only take one list of either Exceptions or ErrorConditions.